gqueue.h

```diff
@@ -30,24 +29,37 @@ static inline void queue_init(volatile void *q, int elt_size, int capacity)
   memset((void*)q, 0x00, q_size);
   gq->item_size = elt_size;
   gq->max_capacity = capacity;
+  gq->head = gq->memory;
+  gq->tail = gq->memory;
+  gq->end  = gq->memory + (gq->max_capacity) * gq->item_size;
```

Comment:
You should rename this `last`.  `end` usually means one past the last valid entry, so the code above where you write to the tail even if `tail == end `looks really wrong.